DNA Methylation introduction section

brockclarke opened the pull request:
final draft of introduction to the DNA methylation section here. There's a nice review on methylation age clocks to potentially cite that will be in press at genome biology likely in a few weeks

agitter commented on content/citation-tags.tsv:
```diff
@@ -96,6 +96,9 @@ Hinton2015_dk	arxiv:1503.02531v1
 Hoff	doi:10.1093/nar/gkp327
 Horton1992_assessment	doi:10.1093/nar/20.16.4331
+Horvath2013	doi:10.1186/gb-2013-14-10-r115
+Houseman2012 doi:10.1186/1471-2105-13-86
```
```suggestion
Houseman2012	doi:10.1186/1471-2105-13-86
```

This should fix the build error

cgreene commented:
> final draft of introduction to the DNA methylation section here. There's a nice review on methylation age clocks to potentially cite that will be in press at genome biology likely in a few weeks

Not sure if you want to update for this, but I'm reminding here in case it has come out in the meantime.

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+
+#### Introduction 
+
+DNA Methylation (DNAm), is the process of adding a methyl group to a cytosine in the context of a CpG dinucleotide. 
```
```suggestion
DNA methylation, is the process of adding a methyl group to a cytosine in the context of a CpG dinucleotide. 
```
Didn't look like DNAm was ever used. I'd remove it unless it's going to be used elsewhere.

brockclarke replied:
agree

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+#### Introduction 
+
+DNA Methylation (DNAm), is the process of adding a methyl group to a cytosine in the context of a CpG dinucleotide. 
+This DNA-level epigenetic modification regulates gene transcription, is critical in development, and alterations to DNA methylation are well-established as contributing to pathophysiology of many diseases including cancers. 
```
☝️ needs a reference - it's common knowledge in some places, but it'd be good to point folks to one or more topical papers (maybe even a few + a nice summary review if you have one) since some of our readers will be ML folks, not biologists.

My impression is also that DNA methylation can also be affected by processes that alter genome accessibility, including transcription, and so it can also provides a readout of a cell's past states. If this isn't totally off base, it's probably good to mention it here too. Doesn't the TP53 status thing fall into this category?

brockclarke replied:
added references here for the basics of DNA methylation biology and methylation in disease. The point about genome accessibility is true in that there is a very high correlation of DNA methylation and histone/chromatin measures for accessibility/inaccessibility, though the relationship with time and cell states has mostly been studied in the context of lineage commitment during development and there is not nearly enough data within specific cell types to trace past cell states yet.

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+
+DNA Methylation (DNAm), is the process of adding a methyl group to a cytosine in the context of a CpG dinucleotide. 
+This DNA-level epigenetic modification regulates gene transcription, is critical in development, and alterations to DNA methylation are well-established as contributing to pathophysiology of many diseases including cancers. 
+Studies of DNA methylation have also shed light on the processes involving initial differentiation of stem cells, aging and pathogenesis in response to environmental exposures. 
```
References would help here, again - over-reference the bio side to help ML people get to relevant literature quickly.

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+This DNA-level epigenetic modification regulates gene transcription, is critical in development, and alterations to DNA methylation are well-established as contributing to pathophysiology of many diseases including cancers. 
+Studies of DNA methylation have also shed light on the processes involving initial differentiation of stem cells, aging and pathogenesis in response to environmental exposures. 
+
+Traditional analytic approaches to DNA methylation data often focus on estimating differential DNA methylation between groups or related with an outcome using linear mixed effects models. 
```
add a ref or two

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+Studies of DNA methylation have also shed light on the processes involving initial differentiation of stem cells, aging and pathogenesis in response to environmental exposures. 
+
+Traditional analytic approaches to DNA methylation data often focus on estimating differential DNA methylation between groups or related with an outcome using linear mixed effects models. 
+In addition, a growing application of DNA methylation measures is to infer cellular or subject phenotypes from samples and either examine the relation of these phenotypes with outcomes or disease states directly, and/or include them in models as covariates. 
```
another ref or two for examples

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+Traditional analytic approaches to DNA methylation data often focus on estimating differential DNA methylation between groups or related with an outcome using linear mixed effects models. 
+In addition, a growing application of DNA methylation measures is to infer cellular or subject phenotypes from samples and either examine the relation of these phenotypes with outcomes or disease states directly, and/or include them in models as covariates. 
+For example, inference of subject age using DNA methylation clock approaches are  established [@tag:Horvath2013], and are starting to be applied to test the relation of biological age with disease risk and outcomes [@tag:Kresovich2019]. 
+Importantly, reference DNA methylation profiles of leukocytes are available to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires. 
```
```suggestion
Different cell types have different DNA methylation profiles.
Reference DNA methylation profiles of leukocytes are available, which makes it feasible to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires. 
```
☝️ missing word at the end of this sentence. I also added a short sentence ot introduce the idea that DNA methylation differs between cell types to help the computational reader along.

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+For example, inference of subject age using DNA methylation clock approaches are  established [@tag:Horvath2013], and are starting to be applied to test the relation of biological age with disease risk and outcomes [@tag:Kresovich2019]. 
+Importantly, reference DNA methylation profiles of leukocytes are available to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires. 
+Cell type inference is important for adjusting for cell-type composition in epigenome-wide association studies [@tag:Teschendorff2017].
+While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation, these methods severely restrict the amount of underlying biology that can be understood and correlated with disease manifestations and phenotypes. When a referenced-based library is not available for use, unsupervised methods are able to estimate these immune profiles [@tag:Titus2017]. 
```
```suggestion
While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation, these methods severely restrict the amount of underlying biology that can be understood and correlated with disease manifestations and phenotypes.
When a referenced-based library is not available for use, unsupervised methods are able to estimate these immune profiles [@tag:Titus2017]. 
```

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+Importantly, reference DNA methylation profiles of leukocytes are available to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires. 
+Cell type inference is important for adjusting for cell-type composition in epigenome-wide association studies [@tag:Teschendorff2017].
+While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation, these methods severely restrict the amount of underlying biology that can be understood and correlated with disease manifestations and phenotypes. When a referenced-based library is not available for use, unsupervised methods are able to estimate these immune profiles [@tag:Titus2017]. 
+Methods that do not rely on these reference libraries [@tag:Houseman2016], are limited by being unable to fully capture the nonlinearity of the methylation data. 
```
```suggestion
However, methods that do not rely on reference libraries [@tag:Houseman2016] are unable to fully capture the nonlinearity of methylation data. 
```
I think I understood this correctly. Check before committing the suggestion 😁

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation, these methods severely restrict the amount of underlying biology that can be understood and correlated with disease manifestations and phenotypes. When a referenced-based library is not available for use, unsupervised methods are able to estimate these immune profiles [@tag:Titus2017]. 
+Methods that do not rely on these reference libraries [@tag:Houseman2016], are limited by being unable to fully capture the nonlinearity of the methylation data. 
+
+There are many promising deep learning approaches that serve to expand the number of CpG sites whose DNA methylation state can be studied by capturing the complex interactions between different regions of DNA and extract the complete set of informative biologically relevant features. 
```
```suggestion
Deep learning approaches have numerous potential applications for DNA methylation data.
Imputation methods that capture complex interactions between different regions of DNA can expand the number of CpG sites whose DNA methylation state can be studied, and ideally these methods can derive their own informative, biologically-relevant features.
```

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,25 @@ Deep learning applied to gene expression data is still in its infancy, but the f
+Methods that do not rely on these reference libraries [@tag:Houseman2016], are limited by being unable to fully capture the nonlinearity of the methylation data. 
+
+There are many promising deep learning approaches that serve to expand the number of CpG sites whose DNA methylation state can be studied by capturing the complex interactions between different regions of DNA and extract the complete set of informative biologically relevant features. 
+The main approaches focus on: 1) estimating regions of methylation status and imputing missing methylation values, 2) performing classification and regression tasks, and 3) understanding latent embeddings of methylation states from which to extract biologically meaningful features, infer interpolated disease states, and uncover relevant CpG sites for the above prediction tasks.
```
```suggestion
The primary deep learning methods developed to date focus on: 1) estimating regions of methylation status and imputing missing methylation values, 2) performing classification and regression tasks, and 3) using the latent embeddings of methylation states to derive biologically meaningful features, infer interpolated disease states, and uncover CpG sites that aid the above prediction tasks.
```

cgreene replied:
See if this keeps the sentiment you were going for.

brockclarke commented:
> > final draft of introduction to the DNA methylation section here. There's a nice review on methylation age clocks to potentially cite that will be in press at genome biology likely in a few weeks
> 
> Not sure if you want to update for this, but I'm reminding here in case it has come out in the meantime.

not out yet unfortunately..

cgreene commented:
I'm guessing you didn't mean to close the PR. Reopening.

cgreene commented on content/04.study.md:
```diff
@@ -47,6 +47,28 @@ Deep learning applied to gene expression data is still in its infancy, but the f
 Many previously untestable hypotheses can now be interrogated as deep learning enables analysis of increasing amounts of data generated by new technologies.
 For example, the effects of cellular heterogeneity on basic biology and disease etiology can now be explored by single-cell RNA-seq and high-throughput fluorescence-based imaging, techniques we discuss below that will benefit immensely from deep learning approaches.
 
+### DNA Methylation
```
```suggestion
### DNA methylation
```

With the change coming in from @agitter ths will line up the headers.

cgreene commented:
I think there are still a few things to correct. I resolved the merge conflicts. Tag me when you are ready for a re-review!

cgreene commented on content/citation-tags.tsv:
```diff
@@ -126,21 +129,21 @@ Koh2016_denoising	doi:10.1101/052118
+Kwabi-Addo2007	doi:10.1158/1078-0432.CCR-07-0085
 Khwaja2017	doi:10.1109/BIOCAS.2017.8325078
-Khwaja2018	arxiv:1810.01243
+Khwaja2018	doi:1810.01243
```
```suggestion
Khwaja2018	arxiv:1810.01243
```
pretty sure this is an arxiv ref. It's why the build is failing

cgreene commented:
Let me know when you want me to take another look - it seems you are addressing the comments now 👍

brockclarke commented:
> Let me know when you want me to take another look - it seems you are addressing the comments now 👍

I have addressed the comments and added refs. etc. you can take a look whenever you can! thanks

cgreene commented on content/04.study.md:
```diff
@@ -49,6 +49,25 @@ For example, the effects of cellular heterogeneity on basic biology and disease
+In addition, a growing application of DNA methylation measures is to infer cellular or subject phenotypes from samples and either examine the relation of these phenotypes with outcomes or disease states directly, and/or include them in models as covariates [@tag:Titus2017], [@tag:Salas2018_GR], [@tag:Zhang2019], [@tag:Horvath2014], [@tag:Quach2017]. 
+For example, inference of subject age using DNA methylation clock approaches are  established [@tag:Horvath2013], and are starting to be applied to test the relation of biological age with disease risk and outcomes [@tag:Kresovich2019]. 
+Different cell types have different DNA methylation profiles.
+Reference DNA methylation profiles of leukocytes are available, which makes it feasible to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires DNA. 
```
```suggestion
A novel approach to immunophenotyping combines measurements with Reference DNA methylation profiles of leukocytes, which are available, to infer immune cell type proportions [@tag:Houseman2012],[@tag:Salas2018].
This strategy is particularly helpful when only DNA is available from a sample.
```

cgreene commented on content/04.study.md:
```diff
@@ -49,6 +49,25 @@ For example, the effects of cellular heterogeneity on basic biology and disease
+Different cell types have different DNA methylation profiles.
+Reference DNA methylation profiles of leukocytes are available, which makes it feasible to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires DNA. 
+Cell type inference is important for adjusting for cell-type composition in epigenome-wide association studies [@tag:Teschendorff2017].
+While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation and has broad utiltiy, cell composition mixtures per se, and methods to incorporate estimates of mixures pose important considerations on the interpretation of underlying biology associated with disease manifestations and phenotypes. 
```
```suggestion
While reference-based libraries have strong predictive value for immune cell type estimation and has broad utility, cell composition mixtures per se, and methods to incorporate estimates of mixtures pose important considerations on the interpretation of underlying biology associated with disease manifestations and phenotypes. 
```

cgreene commented on content/04.study.md:
```diff
@@ -49,6 +49,25 @@ For example, the effects of cellular heterogeneity on basic biology and disease
+Reference DNA methylation profiles of leukocytes are available, which makes it feasible to infer immune cell type proportions with measures of DNA methylation in peripheral blood DNA [@tag:Houseman2012],[@tag:Salas2018], a novel approach to immunophenotyping that only requires DNA. 
+Cell type inference is important for adjusting for cell-type composition in epigenome-wide association studies [@tag:Teschendorff2017].
+While utilizing reference-based libraries demonstrates strong predictive value for immune cell type estimation and has broad utiltiy, cell composition mixtures per se, and methods to incorporate estimates of mixures pose important considerations on the interpretation of underlying biology associated with disease manifestations and phenotypes. 
+When a referenced-based library is not available for use, methods that do not rely on these reference libraries, so called reference-free deconvolution [@tag:Houseman2016], are available to decompose singal purported to be contributed by cell types.
```
```suggestion
When a referenced library is not available for use, methods that do not rely on these reference libraries, so called reference-free deconvolution [@tag:Houseman2016], are available to decompose signal purported to be contributed by cell types.
```

agitter commented on content/citation-tags.tsv:
```diff
@@ -71,8 +69,10 @@ Esfahani2016_melanoma	doi:10.1109/EMBC.2016.7590963
+Feinberg2018	doi:10.1056/NEJMra1402513
 Finnegan2017_maximum	doi:10.1101/105957
 Fong2017_perturb	doi:10.1109/ICCV.2017.371
+Fraga2005	doi:doi.org/10.1073/pnas.0500398102
```
```suggestion
Fraga2005	doi:10.1073/pnas.0500398102
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -10,7 +10,6 @@ Asgari	doi:10.1371/journal.pone.0141287
 blast	doi:10.1016/S0022-2836(05)80360-2
 Angermueller2016_dl_review	doi:10.15252/msb.20156651
 Angermueller2016_single_methyl	doi:10.1186/s13059-017-1189-z
```
```suggestion
Angermueller2016_single_methyl	doi:10.1186/s13059-017-1189-z
Angermueller2017	doi:10.1186/s13059-017-1189-z
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -28,7 +27,6 @@ Bracken2016_mirna	doi:10.1038/nrg.2016.134
 Boza	doi:10.1371/journal.pone.0178751
 Buggenthin2017_imaged_lineage	doi:10.1038/nmeth.4182
 Burlina2016_amd	doi:10.1109/ISBI.2016.7493240
```
```suggestion
Burlina2016_amd	doi:10.1109/ISBI.2016.7493240
Chatterjee2018	arxiv:1807.09617
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -41,14 +39,14 @@ Chen2015_trans_species	doi:10.1093/bioinformatics/btv315
 Codella2016_ensemble_melanoma	arxiv:1610.04662
 Consortium2012_encode	doi:10.1038/nature11247
 CudNN	arxiv:1410.0759
 Dahl2014_multi_qsar	arxiv:1406.1231
```
```suggestion
Dahl2014_multi_qsar	arxiv:1406.1231
Darst2018	doi:10.1186/s12863-018-0646-3
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -100,11 +100,14 @@ Hinton2015_dk	arxiv:1503.02531v1
 Hubara2016_qnn	arxiv:1609.07061
 Huddar2016_predicting	doi:10.1109/ACCESS.2016.2618775
 Hughes2016_macromol_react	doi:10.1021/acscentsci.6b00162
 Iglovikov2017_baa	doi:10.1101/234120
```
```suggestion
Iglovikov2017_baa	doi:10.1101/234120
Islam2018	doi:10.1186/s12919-018-0121-1
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -126,21 +129,21 @@ Koh2016_denoising	doi:10.1101/052118
 Koh2017_understanding	arxiv:1703.04730
 Kooi2016_mamm_lesions	doi:10.1016/j.media.2016.07.007
 Kooi2017_mamm_tl	doi:10.1002/mp.12110
```
```suggestion
Kooi2017_mamm_tl	doi:10.1002/mp.12110
Korfiatis2017	doi:10.1007/s10278-017-0009-z
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -126,21 +129,21 @@ Koh2016_denoising	doi:10.1101/052118
 Lee2016_deeptarget	arxiv:1603.09123v2
 Lee2016_emr_oct_amd	doi:10.1101/094276
 Lei2016_rationalizing	arxiv:1606.04155
 Leibig2016_dr	doi:10.1101/084210
```
```suggestion
Leibig2016_dr	doi:10.1101/084210
Levy2019	doi:10.1101/692665
Levy-Jurgenson2018	doi:10.1101/491357
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -169,18 +173,18 @@ Mapreduce	doi:10.1145/1327452.1327492
+Meissner2008	doi:10.1038/nature07107
 Metaphlan	doi:10.1038/nmeth.2066
 Meng2016_mllib	arxiv:1505.06807
 Min2016_deepenhancer	doi:10.1109/BIBM.2016.7822593
```
```suggestion
Min2016_deepenhancer	doi:10.1109/BIBM.2016.7822593
Momeni2018	doi:10.1101/438341
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -169,18 +173,18 @@ Mapreduce	doi:10.1145/1327452.1327492
 nbc	doi:10.1093/bioinformatics/btq619
 Murdoch2017_automatic	arxiv:1702.02540
+Nazor2012	doi:10.1016/j.stem.2012.02.013
 Nemati2016_rl	doi:10.1109/EMBC.2016.7591355
```
```suggestion
Nemati2016_rl	doi:10.1109/EMBC.2016.7591355
Ni2018	doi:10.1101/385849
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -192,18 +196,16 @@ onecodex	url:https://www.onecodex.com/
 Papernot2017_pate	url:https://openreview.net/forum?id=HkwoSDPgg
 Park2016_deepmirgene	arxiv:1605.00017
 Parnamaa2017	doi:10.1534/g3.116.033654
```
```suggestion
Parnamaa2017	doi:10.1534/g3.116.033654
Pan2018	doi:10.1101/438218
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -192,18 +196,16 @@ onecodex	url:https://www.onecodex.com/
 Park2016_deepmirgene	arxiv:1605.00017
 Parnamaa2017	doi:10.1534/g3.116.033654
-Pan2018	doi:10.1101/438218
 Pawlowski2016	doi:10.1101/085118
```
```suggestion
Pawlowski2016	doi:10.1101/085118
Peng2019	doi:10.1101/527044
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -192,18 +196,16 @@ onecodex	url:https://www.onecodex.com/
+Quach2017	doi:10.18632/aging.101168
 Quang2017_factor	doi:10.1101/151274
 Qin2017_onehot	doi:10.1371/journal.pcbi.1005403
 Qiu2017_graph_embedding	doi:10.1101/110668
```
```suggestion
Qiu2017_graph_embedding	doi:10.1101/110668
Qiu2018	doi:10.1101/406066
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -261,7 +267,8 @@ Tan2015_adage	doi:10.1128/mSystems.00025-15
 TAC-ELM	doi:10.1142/S0219720012500151
 TensorFlow	arxiv:1603.04467
-Tian2019	doi:10.1186/s12864-019-5488-5
+Teschendorff2017	doi:10.2217/epi-2016-0153
```
```suggestion
Teschendorff2017	doi:10.2217/epi-2016-0153
Tian2019	doi:10.1186/s12864-019-5488-5
```

cgreene commented on content/citation-tags.tsv:
```diff
@@ -271,11 +278,11 @@ Vera2016_sc_analysis	doi:10.1146/annurev-genet-120215-034854
 Vervier	doi:10.1093/bioinformatics/btv683
 Wallach2015_atom_net	arxiv:1510.02855
 Wang2016_breast_cancer	arxiv:1606.05718
```
```suggestion
Wang2016_breast_cancer	arxiv:1606.05718
Wang2016_methyl	doi:10.1038/srep19598
```

agitter commented:
The build failed because there are missing tag definitions:
```
## ERROR
CSL JSON Data retrieval failed for:
tag:Levy-Jurgenson2018
tag:Levy2019
tag:Momeni2018
tag:Ni2018
tag:Roberston2005
```

The tag file does have `Robertson2005`, is that what `tag:Roberston2005` should match?

brockclarke commented:
> The build failed because there are missing tag definitions:
> 
> ```
> ## ERROR
> CSL JSON Data retrieval failed for:
> tag:Levy-Jurgenson2018
> tag:Levy2019
> tag:Momeni2018
> tag:Ni2018
> tag:Roberston2005
> ```
> 
> The tag file does have `Robertson2005`, is that what `tag:Roberston2005` should match?

Robertson2005	doi:10.1038/nrg1655
pushed spelling error correction

brockclarke commented:
> > The build failed because there are missing tag definitions:
> > ```
> > ## ERROR
> > CSL JSON Data retrieval failed for:
> > tag:Levy-Jurgenson2018
> > tag:Levy2019
> > tag:Momeni2018
> > tag:Ni2018
> > tag:Roberston2005
> > ```
> > 
> > 
> > The tag file does have `Robertson2005`, is that what `tag:Roberston2005` should match?
> 
> Robertson2005	doi:10.1038/nrg1655
> pushed spelling error correction

also, > > tag:Levy-Jurgenson2018
> > tag:Levy2019
> > tag:Momeni2018
> > tag:Ni2018 are in jlevy44's section

agitter commented:
@brockclarke can you please commit the rest of @cgreene's suggestions?  We don't have the ability to directly commit to this pull request branch.  His edits should fix the rest of the deleted tags.

agitter commented:
Please disregard my detailed comments on specific lines above.  I was going to make style suggestions in this pull request but now think that should be done separately.
